src/Timer/Timer.tsx

```diff
@@ -1,83 +1,63 @@
 import React from "react";
-
-export interface TimerProps {
-    callback?: VoidFunction;
-    className?: string;
-    duration: number;
-    id?: string;
-}
-
-interface TimerState {
-    timer: string;
-}
-
-export class Timer extends React.Component<TimerProps, TimerState> {
-    private innerInterval: any;
-    constructor(props: TimerProps) {
-        super(props);
-
-        this.state = {
-            timer: "00:00",
-        };
-    }
-
-    startInterval(timeout: number): void {
-        this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-            this.clearInterval();
-            this.innerInterval = setInterval(() => {
-                if (timeout > 0) {
-                    timeout = timeout - 1000;
-                    this.setState({ timer: this.convertMStoTime(timeout) }, () => {
-                        if (timeout === 0) {
-                            this.props.callback();
-                            this.clearInterval();
-                        }
-                    });
-                }
-            }, 1000);
-        });
-    }
-
-    clearInterval(): void {
-        if (this.innerInterval) {
-            clearTimeout(this.innerInterval);
-            this.innerInterval = null;
-        }
-    }
-
-    convertMStoTime(value: number): string {
-        const date: Date = new Date(value);
-        return (
-            (date.getUTCHours() > 0 ? date.getUTCHours() + ":" : "") +
-            (date.getUTCMinutes() < 10 ? "0" + date.getUTCMinutes() : date.getUTCMinutes()) +
-            ":" +
-            (date.getUTCSeconds() < 10 ? "0" + date.getUTCSeconds() : date.getUTCSeconds())
-        );
-    }
-
-    componentDidMount() {
-        if (this.props.duration !== null && this.props.duration !== undefined) {
-            this.startInterval(this.props.duration);
-        }
-    }
-
-    componentWillUnmount() {
-        this.clearInterval();
-    }
-
-    componentDidUpdate(prevProps: TimerProps): void {
-        if (prevProps.duration !== this.props.duration) {
-            if (this.props.duration !== null && this.props.duration !== undefined) {
-                this.startInterval(this.props.duration);
+import classnames from "classnames";
+import { min } from "lodash";
+
+export type TimerProps = JSX.IntrinsicElements["time"] & {
+    duration?: number;
+    onTimerEnded?: () => void;
+};
+
+export const Timer: React.FC<TimerProps> = React.memo(({ duration, onTimerEnded, ...props }: TimerProps) => {
+    const [time, setTime] = React.useState<string>("00:00");
+    const countDown = React.useRef<number>(0);
+    const timer = React.useRef<any>(0);
```

Comment:
Is there any reason while the type here is any ?

Reply:
The returned type when setting `setInterval` or `setTimeout` is a number. Typescript will complain about that because it thinks it's of type `NodeJS.Timeout` or `NodeJS.Timer` respectively. Using either of them will throw an error during build. Therefore, I always use `any` for intervals and timeouts.